### PR TITLE
`Discussion`: Filter posts with CriteriaBuilder

### PR DESCRIPTION
**Pull request by canberkanar**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [ ] I implemented the changes with a good performance and prevented too many database calls.
- [ ] I documented the Java code using JavaDoc style.
#### Client
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [ ] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java [hunk 1]
```diff
@@ -1,25 +1,45 @@
...
+                                .and(getCourseWideContextSpecification(postContextFilter.getCourseWideContext()))
+                                .and(getOwnSpecification(postContextFilter.getFilterToOwn(), userId)))
+                        .and(getAnsweredOrReactedSpecification(postContextFilter.getFilterToAnsweredOrReacted(), userId))
+                        .and(getUnresolvedSpecification(postContextFilter.getFilterToUnresolved())));
+        // .and(getSortSpecification(postContextFilter.getPostSortCriterion(), postContextFilter.getSortingOrder())));
+
+        return findAll(specification);
+    }
+
+    List<Post> findAll(Specification<Post> spec);
```
Comment (Dominik-Weinzierl) [inline src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java:41]:
```suggestion
```
This is not necessary, as it should already be present (JpaSpecificationExecutor)

File: src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import de.tum.in.www1.artemis.domain.enumeration.DisplayPriority;
+import de.tum.in.www1.artemis.domain.enumeration.SortingOrder;
+import de.tum.in.www1.artemis.domain.metis.*;
+
+public class PostSpecs {
+
+    public static Specification<Post> getCourseSpecification(Long courseId, Long lectureId, Long exerciseId) {
+        return (root, query, criteriaBuilder) -> {
+            if (lectureId != null || exerciseId != null) {
+                return criteriaBuilder.conjunction();
```
Comment (Dominik-Weinzierl) [inline src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:20]: You can just return null :+1:

File: src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.PINNED)), 1)
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.NONE)), 2)
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.ARCHIVED)), 3);
+            orderList.add(criteriaBuilder.asc(caseExpression));
+
+            if (postSortCriterion != null && postSortCriterion != PostSortCriterion.CREATION_DATE) {
+                orderList.add(sortingOrder == SortingOrder.ASCENDING ? criteriaBuilder.asc(root.get(Post_.CREATION_DATE)) : criteriaBuilder.desc(root.get(Post_.CREATION_DATE)));
+            }
+
+            query.orderBy(orderList);
```
Comment (Dominik-Weinzierl) [inline src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:139]: Can this be done as we did in `UserRepository.java -> getAllManagedUsers` :thinking:

e.g. similar to:
```
var sorting = Sort.by(userSearch.getSortedColumn());
sorting = userSearch.getSortingOrder() == SortingOrder.ASCENDING ? sorting.ascending() : sorting.descending();
final var sorted = PageRequest.of(userSearch.getPage(), userSearch.getPageSize(), sorting);

...

return findAll(..., sorted);
```

**Review by Dominik-Weinzierl — COMMENTED** [review]
I know this is still a draft but i have some suggestions and questions :D

File: src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java (hunk 2 above)
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:20]: I'm not sure why but in some cases it throws an error that way.

Comment (Dominik-Weinzierl) [reply src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:20]: mhm good to know :thinking:

File: src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java (hunk 3 above)
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:139]: This is just a draft. In fact I've not completed the sorting part yet. 😆

File: src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java (hunk 1 above)
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java:41]: Nice one, thanks 👍🏻

File: src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java (hunk 3 above)
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java:139]: No unfortunately. We have a lot of variations. We can't just sort by column. In some cases we have to sort by number of answers or number of a specific type of emoji. I'll take a look at the sorting part after I'm done with the filters though. This part very preliminary. (It is just a draft at the moment 🙄 )

**Review by jbllhff — APPROVED** [review]
Tested on ts2, works in general, one minor comments:
It seems like filtering for "Random" posts does not work in course discussion, no results were found there.

**Comment by Dominik-Weinzierl** [comment]
@canberkanar Can you add the generated queries as dropdown into the description? I guess this would be really nice :+1:

**Comment by canberkanar** [comment]
> @canberkanar Can you add the generated queries as dropdown into the description? I guess this would be really nice 👍

Added them to the PR description. 😃

**Comment by babyygemperor** [comment]
Tested on ts1 in a testing session. Just like [Lukas](https://github.com/ls1intum/Artemis/pull/5257#pullrequestreview-1019977255) mentioned, same issue, but when both `unresolved` and `reacted` is selected, the amount of posts showing get squared.
E.g. - 4 with just reaction or unresolved, and 16 with both.
image

**Review by minrows — COMMENTED** [review]
Tested on ts1 during the testing session. 
Found a bug, I was not able to filter the posts based on my search string.

**Comment by Santia-go** [comment]
If I have reacted 2 times in the same post, when I filter it, I can see the post 2 times. 
image

**Comment by canberkanar** [comment]
> Tested on ts1 in a testing session. Just like [Lukas](https://github.com/ls1intum/Artemis/pull/5257#pullrequestreview-1019977255) mentioned, same issue, but when both `unresolved` and `reacted` is selected, the amount of posts showing get squared. E.g. - 4 with just reaction or unresolved, and 16 with both. image

Issue fixed. Please take another look.

**Comment by canberkanar** [comment]
> Tested on ts1 during the testing session. Found a bug, I was not able to filter the posts based on my search string.

Issue fixed. Please take another look.

**Comment by canberkanar** [comment]
Server tests for Post ordering to be fixed

**Review by ole-ve — APPROVED** [review]
Tested during the testing session, everything worked as expected 👍

**Review by xlieA — APPROVED** [review]
Tested during testing session. Works as described.

**Review by minrows — DISMISSED** [review]
Tested on ts1 during testing session.
Everything worked as explained and the previous issue of not being able to search the string is also working fine. 👍

**Review by akesfeden — DISMISSED** [review]
Tested on TS1 during testing session, filtering and sorting works smoothly and as expected.

**Review by jschmautzer — DISMISSED** [review]
tested on ts2, filtering and sorting works as described

**Review by Dominik-Weinzierl — COMMENTED** [review]
Some suggestions/questions and it seems like there are some fixes available :thinking:

**Review by babyygemperor — DISMISSED** [review]
Tested on ts1, works as described

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on ts1.
Tested with course, lecture and exercise. Works as intended for all of them.

**Review by alaraozdenler — COMMENTED** [review]
Tested on ts2, worked as described on Course Discussion Page. However, I could not reproduce the 4th criterion on Exercise and Lecture Discussion Tabs. Even though I changed the vote-emoji count, the order of the posts stayed the same.

**Comment by canberkanar** [comment]
> Tested on ts2, worked as described on Course Discussion Page. However, I could not reproduce the 4th criterion on Exercise and Lecture Discussion Tabs. Even though I changed the vote-emoji count, the order of the posts stayed the same.

Thank you for your feedback. Did you test the page done like in the following screencast? Sometimes written instructions may be confusing and hard to follow. 😢
[sorting.webm](https://user-images.githubusercontent.com/67344158/177113637-aa48548c-3d62-4285-af18-1b006e819f2d.webm)

**Comment by alaraozdenler** [comment]
> > Tested on ts2, worked as described on Course Discussion Page. However, I could not reproduce the 4th criterion on Exercise and Lecture Discussion Tabs. Even though I changed the vote-emoji count, the order of the posts stayed the same.
> 
> Thank you for your feedback. Did you test the page done like in the following screencast? Sometimes written instructions may be confusing and hard to follow. 😢 [sorting.webm](https://user-images.githubusercontent.com/67344158/177113637-aa48548c-3d62-4285-af18-1b006e819f2d.webm)


Yes, I tested it as the screencast showed.

**Comment by canberkanar** [comment]
> > > Tested on ts2, worked as described on Course Discussion Page. However, I could not reproduce the 4th criterion on Exercise and Lecture Discussion Tabs. Even though I changed the vote-emoji count, the order of the posts stayed the same.
> > 
> > 
> > Thank you for your feedback. Did you test the page done like in the following screencast? Sometimes written instructions may be confusing and hard to follow. 😢 [sorting.webm](https://user-images.githubusercontent.com/67344158/177113637-aa48548c-3d62-4285-af18-1b006e819f2d.webm)
> 
> Yes, I tested it as the screencast showed.

I see. If testing was done with 'Default Sorting' and posts are not ordered by vote count as expected, please attach a screencast so I can take a look in detail. Thanks 😊
